lib/gems/pending/util/postgres_admin.rb

```diff
@@ -136,12 +136,26 @@ def self.backup_pg_compress(opts)
     opts.delete(:dbname)
 
     path = Pathname.new(opts.delete(:local_file))
-
     FileUtils.mkdir_p(path.dirname)
-    Dir.mktmpdir("vmdb_backup", path.dirname) do |dir|
-      runcmd("pg_basebackup", opts, :z => nil, :format => "t", :xlog_method => "fetch", :pgdata => dir)
-      FileUtils.mv(File.join(dir, "base.tar.gz"), path)
-    end
+
+    # Build commandline from AwesomeSpawn
+    args = {:z => nil, :format => "t", :xlog_method => "fetch", :pgdata => "-"}
+    cmd  = AwesomeSpawn.build_command_line("pg_basebackup", combine_command_args(opts, args))
+    $log.info("MIQ(#{name}.#{__method__}) Running command... #{cmd}")
+
+    # Run command in a separate thread
+    read, write   = IO.pipe
+    error_path    = Dir::Tmpname.create("") { |tmpname| tmpname }
+    process       = Process.detach(Kernel.spawn(pg_env(opts), cmd, :out => write, :err => error_path))
+    stream_reader = Thread.new { IO.copy_stream(read, path) } # Copy output to path
+    write.close
+
+    # Wait for them to finish
+    pid = process.value
```

Comment:
Is this behavior documented anywhere? At this point `process` is a thread object and `#value` essentially does `#join` and returns the ["value" of the thread](https://ruby-doc.org/core-2.4.1/Thread.html#method-i-value) in this case are we sure that the thread that is created by `detach` will return something that responds to `#exitstatus`? I don't see that documented anywhere.

Reply:
I guess I'm still just confused why we can't `Process.wait` in the main thread.

Reply:
> I guess I'm still just confused why we can't Process.wait in the main thread.

I think I covered most of this here (or at least the "why" part of your question):

https://github.com/ManageIQ/manageiq-gems-pending/pull/365/files#r213669910

Basically, to use vanilla `Process.wait`, I would be doing:

```ruby
Process.wait(pid)
status = Process::Status.new(pid).exitstatus
```

> Is this behavior documented anywhere?

Actually, I am having trouble finding it, but I swear I found it somewhere directly...

Anyway, where I did find it's usage of `Process.detach` is in `Open3`, in both the [`#popen_run`](https://github.com/ruby/ruby/blob/ac543abe/lib/open3.rb#L200) and [`#pipeline_run`](https://github.com/ruby/ruby/blob/ac543abe/lib/open3.rb#L659) methods, and it's recommendation to use `.value` in basically all of it's documentation:

https://ruby-doc.org/stdlib-1.9.3/libdoc/open3/rdoc/Open3.html#method-c-popen3

So I am not pulling this out of no where.

Reply:
:+1: Can we just use a better variable name than `pid` for this? Makes it seem like it's an integer then it looks confusing when we call `#exitstatus` on it.

Reply:
Right on, yeah, I am fine with that.

For the above, I was thinking of `status`, and for your other comment `process_thread`.  So:

```
pid     -> status
process -> process_thread
```

Sound reasonable?